### PR TITLE
feat(broker): support error boundary event

### LINKED ISSUE #3497: Catch an error on a boundary event
**Is your feature request related to a problem? Please describe.**
A job worker can throw an error when a business error (i.e. non-technical) occurs and the job can not be completed. This error should be handled in the workflow by catching it on a boundary event.

**Describe the solution you'd like**
* the error event contains an `errorCode` (any string)
* the error is caught on the boundary event if the error code in the BPMN is equal to the thrown error code
* the boundary can be attached to the corresponding service task or on an upper scope

**Describe alternatives you've considered**
none

**Additional context**
none

### PR DESCRIPTION
**Pull request by saig0**
## Description

* deploy workflow with error boundary events
* new job intents: THROW_ERROR, ERROR_THROWN
* accept job throw error command if job is activatable
* trigger error boundary event if error code matches
* update job metrics when an error is thrown

:warning: BREAKING CHANGES :warning: 
* add 'errorCode' property to job and job batch JSON serialization format
* add 'errorCode' to Elasticsearch job and job batch template

## Related issues

closes #3497 

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

File: engine/src/main/java/io/zeebe/engine/processor/workflow/job/JobErrorThrownProcessor.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    final var job = record.getValue();
+    final var serviceTaskInstanceKey = job.getElementInstanceKey();
+    final var serviceTaskInstance = elementInstanceState.getInstance(serviceTaskInstanceKey);
+
+    if (serviceTaskInstance != null && serviceTaskInstance.isActive()) {
+
+      final var errorCode = job.getErrorCodeBuffer();
+      final var workflow = getWorkflow(job.getWorkflowKey());
+
+      final var foundCatchEvent = findCatchEvent(workflow, serviceTaskInstance, errorCode);
```
Comment (MiguelPires) [inline engine/src/main/java/io/zeebe/engine/processor/workflow/job/JobErrorThrownProcessor.java]: Personally, I don't like this pattern of setting a field in the function but then returning it anyway. Usually the variable then shadows the field and then it forces me to see if they're actually the same object and if not then the variable name should be changed. I get that it removes the allocation from the hot path but, as you said in a comment bellow, we're assuming errors happen rarely so it wouldn't be a big deal to have it in `findCatchEvent`. These are just my 2 cents and it's something we do frequently, so I'm fine with keeping it. It makes me wonder however, if it makes sense to keep the "Local variables should not shadow class fields" code smell on SonarCloud.

**Review by miguelpires — APPROVED** [review]
I've just a couple of small comments on readibility but they're not crucial so I'm approving it anyway

File: bpmn-model/src/main/java/io/zeebe/model/bpmn/util/ModelUtil.java [hunk 2]
```diff
@@ -27,13 +28,14 @@
...
+
+  public static <E extends EventDefinition> Stream<E> getBoundaryEvents(
+      final Activity activity, final Class<E> definitionType) {
 
     return activity.getBoundaryEvents().stream()
         .flatMap(event -> event.getEventDefinitions().stream())
-        .filter(definition -> definition instanceof MessageEventDefinition)
-        .map(MessageEventDefinition.class::cast);
+        .filter(definition -> definitionType.isAssignableFrom(definition.getClass()))
+        .map(definition -> definitionType.cast(definition));
```
[the remark this answers is not in the document]
Comment (saig0) [reply bpmn-model/src/main/java/io/zeebe/model/bpmn/util/ModelUtil.java]: True :+1:

File: bpmn-model/src/main/java/io/zeebe/model/bpmn/validation/zeebe/ActivityValidator.java [hunk 3]
```diff
@@ -45,4 +59,27 @@ public void validate(
...
+            .filter(e -> e.getError() != null)
+            .map(ErrorEventDefinition::getError)
+            .filter(error -> error.getErrorCode() != null && !error.getErrorCode().isEmpty())
+            .collect(groupingBy(Error::getErrorCode, counting()));
+
+    presenceByErrorCode.entrySet().stream()
+        .filter(e -> e.getValue() > 1)
+        .map(Entry::getKey)
+        .forEach(
+            errorCode -> {
```
[the remark this answers is not in the document]
Comment (saig0) [reply bpmn-model/src/main/java/io/zeebe/model/bpmn/validation/zeebe/ActivityValidator.java]: :+1:

[the remark this answers is not in the document]
Comment (saig0) [reply bpmn-model/src/main/java/io/zeebe/model/bpmn/validation/zeebe/ActivityValidator.java]: :+1:

File: engine/src/main/java/io/zeebe/engine/processor/workflow/job/JobErrorThrownProcessor.java (hunk 1 above)
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/job/JobErrorThrownProcessor.java]: Agree. I also don't like this style but I'm not sure how to make it better - if it should be allocation free. I renamed the field to not shadow the local variable.

**Comment by saig0** [comment]
@MiguelPires, thank you for the review :+1: I addressed your comments.

**Pull request merged**